pygfx/objects/_events.py

```diff
@@ -0,0 +1,358 @@
+        super().__init__(*args, **kwargs)
+        self.dx = dx
+        self.dy = dy
+        self.x = x
+        self.y = y
+        self.modifiers = modifiers or ()
+
+
+class WindowEvent(Event):
+    def __init__(self, *args, width=None, height=None, pixel_ratio=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.width = width
+        self.height = height
+        self.pixel_ratio = pixel_ratio
+
+
+class EventTarget:
+    """Mixin class that enables event handlers to be attached to objects
+    of the mixed-in class.
+    """
+
+    pointer_captures = WeakValueDictionary()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._event_handlers = defaultdict(set)
+
+    def add_event_handler(self, *args):
+        """Register an event handler.
+
+        Arguments:
+            callback (callable): The event handler. Must accept a
+                single event argument.
+            *types (list of strings): A list of event types.
+
+        For the available event types, see
+        https://jupyter-rfb.readthedocs.io/en/latest/events.html
+
+        Can also be used as a decorator.
+
+        Example:
+
+        .. code-block:: py
+
+            def my_handler(event):
+                print(event)
+
+            obj.add_event_handler(my_handler, "pointer_up", "pointer_down")
+
+        Decorator usage example:
+
+        .. code-block:: py
+
+            @obj.add_event_handler("pointer_up", "pointer_down")
+            def my_handler(event):
+                print(event)
+        """
+
+        decorating = not callable(args[0])
+        callback = None if decorating else args[0]
+        types = args if decorating else args[1:]
+
+        if not types:
+            raise ValueError("No types registered for callback")
+
+        def decorator(_callback):
+            for type in types:
+                self._event_handlers[type].add(_callback)
+            return _callback
+
+        if decorating:
+            return decorator
+        return decorator(callback)
+
+    def remove_event_handler(self, callback, *types):
+        """Unregister an event handler.
+
+        Arguments:
+            callback (callable): The event handler.
+            *types (list of strings): A list of event types.
+        """
+        for type in types:
+            self._event_handlers[type].remove(callback)
+
+    def handle_event(self, event: Event):
+        """Handle an incoming event.
+
+        Subclasses can overload this method. Events include widget
+        resize, mouse/touch interaction, key events, and more. An event
+        is a dict with at least the key event_type. For details, see
+        https://jupyter-rfb.readthedocs.io/en/latest/events.html
+        """
+        event_type = event.type
+        for callback in self._event_handlers[event_type]:
+            if event.cancelled:
+                break
+            callback(event)
```

Comment:
```suggestion
            try:
                callback(event)
            except Exception:
                log_exception(f"Error during handling {event_type} event", err)
```

And you'd need this, taken from wgpu-py.
```py
import logging

logger = logging.getLogger("pygfx")

err_hashes = {}

def log_exception(kind, err):
    """Log the given exception instance, but only log a one-liner for
    subsequent occurances of the same error to avoid spamming (which
    can happen easily with errors in the event handlers).
    """
    msg = str(err)
    msgh = hash(msg)
    if msgh not in err_hashes:
        # Provide the exception, so the default logger prints a stacktrace.
        # IDE's can get the exception from the root logger for PM debugging.
        err_hashes[msgh] = 1
        logger.error(kind, exc_info=err)
    else:
        # We've seen this message before, return a one-liner instead.
        err_hashes[msgh] = count = err_hashes[msgh] + 1
        msg = kind + ": " + msg.split("\n")[0].strip()
        msg = msg if len(msg) <= 70 else msg[:69] + "…"
        logger.error(msg + f" ({count})")
```

Reply:
mmm ... with a minor change in wgpu-py, we might re-use its `log_exception` function ...

I could pick this up in a later PR.

Reply:
Great suggestion, I'll leave it up to you! I'll resolve this comment for now then.

Reply:
Actually, I'll put this in place already, and then you can update / rework this part later.